trie-db/src/iterator.rs

```diff
@@ -315,6 +321,241 @@ impl<'a, 'cache, L: TrieLayout> TrieDBNodeIterator<'a, 'cache, L> {
+							},
+							Err(err) => {
+								crumb.increment();
+								return Some(Err(err))
+							},
+						}
+					} else {
+						crumb.increment();
+					}
+				},
+				_ => panic!(
+					"Crumb::increment and TrieDBNodeIterator are implemented so that \
+						the above arms are the only possible states"
+				),
+			}
+		}
+	}
+
+	/// Fetches the next trie item.
+	///
+	/// Must be called with the same `db` as when the iterator was created.
+	pub fn next_item(&mut self, db: &TrieDB<L>) -> Option<TrieItem<TrieHash<L>, CError<L>>> {
+		while let Some(raw_item) = self.next_raw_item(db) {
+			let (prefix, _, node) = match raw_item {
+				Ok(raw_item) => raw_item,
+				Err(err) => return Some(Err(err)),
+			};
+
+			let mut prefix = prefix.clone();
+			let value = match node.node() {
+				Node::Leaf(partial, value) => {
+					prefix.append_partial(partial.right());
+					value
+				},
+				Node::Branch(_, value) => match value {
+					Some(value) => value,
+					None => continue,
+				},
+				Node::NibbledBranch(partial, _, value) => {
+					prefix.append_partial(partial.right());
+					match value {
+						Some(value) => value,
+						None => continue,
+					}
+				},
+				_ => continue,
+			};
+
+			let (key_slice, maybe_extra_nibble) = prefix.as_prefix();
+			let key = key_slice.to_vec();
+			if let Some(extra_nibble) = maybe_extra_nibble {
+				return Some(Err(Box::new(TrieError::ValueAtIncompleteKey(key, extra_nibble))))
+			}
+
+			let value = match value {
+				Value::Node(hash) => match Self::fetch_value(db, &hash, (key_slice, None)) {
+					Ok(value) => value,
+					Err(err) => return Some(Err(err)),
+				},
+				Value::Inline(value) => value.to_vec(),
+			};
+
+			return Some(Ok((key, value)))
+		}
+		None
+	}
+
+	/// Fetches the next key.
+	///
+	/// Must be called with the same `db` as when the iterator was created.
+	pub fn next_key(&mut self, db: &TrieDB<L>) -> Option<TrieKeyItem<TrieHash<L>, CError<L>>> {
+		while let Some(raw_item) = self.next_raw_item(db) {
+			let (prefix, _, node) = match raw_item {
+				Ok(raw_item) => raw_item,
+				Err(err) => return Some(Err(err)),
+			};
+
+			let mut prefix = prefix.clone();
+			match node.node() {
+				Node::Leaf(partial, _) => {
+					prefix.append_partial(partial.right());
+				},
+				Node::Branch(_, value) =>
+					if value.is_none() {
+						continue
+					},
+				Node::NibbledBranch(partial, _, value) => {
+					prefix.append_partial(partial.right());
+					if value.is_none() {
+						continue
+					}
+				},
+				_ => continue,
+			};
+
+			let (key_slice, maybe_extra_nibble) = prefix.as_prefix();
+			let key = key_slice.to_vec();
+			if let Some(extra_nibble) = maybe_extra_nibble {
+				return Some(Err(Box::new(TrieError::ValueAtIncompleteKey(key, extra_nibble))))
+			}
+
+			return Some(Ok(key))
+		}
+		None
+	}
+}
+
+/// Iterator for going through all nodes in the trie in pre-order traversal order.
+pub struct TrieDBNodeIterator<'a, 'cache, L: TrieLayout> {
```

Comment:
(Just curious about the design decision) why you decided to separate the `TrieDbNodeIterator` and `TriedDbRawIterator` and not embed everything in one single structure?

Intuition suggests that in this way you can export the raw iterator (`into_raw`) and reuse it to construct another iterator (`from_raw`). Maybe also you've preferred to keep the lower level component separated to have a more clean design?